libs/utils/src/http/endpoint.rs

```diff
@@ -54,60 +58,58 @@ struct RequestId(String);
 /// tries to achive with its `.instrument` used in the current approach.
 ///
 /// If needed, a declarative macro to substitute the |r| ... closure boilerplate could be introduced.
-pub struct RequestSpan<E, R, H>(pub H)
-where
-    E: Into<Box<dyn std::error::Error + Send + Sync>> + 'static,
-    R: Future<Output = Result<Response<Body>, E>> + Send + 'static,
-    H: Fn(Request<Body>) -> R + Send + Sync + 'static;
-
-impl<E, R, H> RequestSpan<E, R, H>
+pub async fn request_span<R, H>(request: Request<Body>, handler: H) -> R::Output
 where
-    E: Into<Box<dyn std::error::Error + Send + Sync>> + 'static,
-    R: Future<Output = Result<Response<Body>, E>> + Send + 'static,
-    H: Fn(Request<Body>) -> R + Send + Sync + 'static,
+    R: Future<Output = Result<Response<Body>, ApiError>> + Send + 'static,
+    H: FnOnce(Request<Body>) -> R + Send + Sync + 'static,
 {
-    /// Creates a tracing span around inner request handler and executes the request handler in the contex of that span.
-    /// Use as `|r| RequestSpan(my_handler).handle(r)` instead of `my_handler` as the request handler to get the span enabled.
-    pub async fn handle(self, request: Request<Body>) -> Result<Response<Body>, E> {
-        let request_id = request.context::<RequestId>().unwrap_or_default().0;
-        let method = request.method();
-        let path = request.uri().path();
-        let request_span = info_span!("request", %method, %path, %request_id);
-
-        let log_quietly = method == Method::GET;
-        async move {
-            let cancellation_guard = RequestCancelled::warn_when_dropped_without_responding();
-            if log_quietly {
-                debug!("Handling request");
-            } else {
-                info!("Handling request");
-            }
+    let request_id = request.context::<RequestId>().unwrap_or_default().0;
+    let method = request.method();
+    let path = request.uri().path();
+    let request_span = info_span!("request", %method, %path, %request_id);
+
+    let log_quietly = method == Method::GET;
+    async move {
+        let cancellation_guard = RequestCancelled::warn_when_dropped_without_responding();
+        if log_quietly {
+            debug!("Handling request");
+        } else {
+            info!("Handling request");
+        }
 
-            // Note that we reuse `error::handler` here and not returning and error at all,
-            // yet cannot use `!` directly in the method signature due to `routerify::RouterBuilder` limitation.
```

Comment:
I didn't understand this comment. What does using `!` mean here? Did it mean `?`, perhaps?

I removed this, but if someone understands what it meant or if we should keep it, I'm all ears.

Reply:
I think because we were unable to use infallible as the err with routerify... I am not sure about **not using `error::handler`**, feels like it's usage was retained on purpose.

what's the upside of not using it?

Reply:
> Or I think the idea here was to use is Result<(), !> if I recall correctly.

A-ha, I see. That makes sense, thanks!

Reply:
I put back a comment to explain that, using words that I understand :-).

Reply:
Earlier thread: https://github.com/neondatabase/neon/pull/3755#discussion_r1130573353

Reply:
My understanding of the earlier thread was that we wanted to continue using it regardless of the possible panic, but didn't really record exploring anything about tightening the error to be `ApiError` always. So I understand it was missed.

It also does not seem plausible that there would had been a handler that would had needed to return a different kind of error, otherwise we would had added it to `error::handler`.

Reply:
Oki, and now I understood on https://github.com/neondatabase/neon/pull/4337#pullrequestreview-1442442050 why `error::handler` is used, it is to be consistent about logging the responses.

While it seems uninteresting it might still make sense to forward just the right lines 105..=112 to a new function, which would be called by this new `request_span` and `error::handler` to log and produce the body.

Reply:
Refactored it that way. I also made `error::handler` a little more lenient, so that it doesn't panic on errors other than ApiError.